File: codes/cpp/polymorphism/overriding.cpp
#include <iostream>
#include <string>

class Person {
public:
    virtual void talk(){ std::cout << "I am a person" << std::endl; }
    void type(){ std::cout << "Person" << std::endl; }
};

class Student: public Person {
public:
    void talk(){ std::cout << "I am a student" << std::endl; }
    void type(){ std::cout << "Student" << std::endl; }
};

int main()
{
    Person *pe = new Person();
    Student *st = new Student();
    Person *pst = st;

    pe->talk(); // I am a person
    st->talk(); // I am a student
    pst->talk(); //I am a student

    pe->type(); //Person
    st->type(); //Student
    pst->type(); //Person

    return 0;
}

File: codes/php/polymorphism/overriding.php
<?php

class Person {
    public function talk(){ echo "I am a person\n"; }
}

class Student extends Person {
    public function talk(){ echo "I am a student\n"; }
}

$pe = new Person();
$st = new Student();
$pe->talk(); // I am a person
$st->talk(); // I am a student

?>
